File: querier/internal/query/dependencies.go
package query

import (
	"context"
	"errors"
	"fmt"
	"path/filepath"
	"sort"
	"strings"

	"dev.research4jar/querier/internal/depgraph"
	"dev.research4jar/querier/internal/project"
)

type DependencyWhyResult struct {
	Coordinate       string   `json:"coordinate"`
	Artifact         string   `json:"artifact"`
	Version          string   `json:"version"`
	Scope            string   `json:"scope,omitempty"`
	Direct           bool     `json:"direct"`
	Depth            int      `json:"depth"`
	Path             []string `json:"path"`
	Parent           string   `json:"parent,omitempty"`
	DirectDependency string   `json:"direct_dependency,omitempty"`
	MatchedBy        string   `json:"matched_by"`
	SourceJar        string   `json:"source_jar,omitempty"`
	SourceClass      string   `json:"source_class,omitempty"`
}

type DependencyWhyResponse struct {
	Query    SymbolRequest         `json:"query"`
	Results  []DependencyWhyResult `json:"results"`
	Total    int                   `json:"total"`
	Coverage Coverage              `json:"coverage"`
}

type manifestSource struct {
	ShardID    string
	Coordinate string
	Filename   string
	Source     string
}

func WhyDependency(
	ctx context.Context,
	pointer project.Pointer,
	manifestPath string,
	projectDir string,
	arg string,
) (DependencyWhyResponse, error) {
	graph, err := depgraph.Load(projectDir)
	if err != nil {
		if errors.Is(err, depgraph.ErrUnsupported) {
			return DependencyWhyResponse{}, errors.New(
				"no dependency provenance found; run research4jar index in a Maven project to create .research4jar/dependencies.json",
			)
		}
		return DependencyWhyResponse{}, fmt.Errorf("load dependency provenance: %w", err)
	}
	sources, err := loadManifestSources(ctx, manifestPath)
	if err != nil {
		return DependencyWhyResponse{}, err
	}

	targets, err := dependencyTargets(ctx, pointer, sources, arg)
	if err != nil {
		return DependencyWhyResponse{}, err
	}
	results := dependencyWhyResults(graph, targets)
	return DependencyWhyResponse{
		Query:    SymbolRequest{Command: "why-dependency", Arg: arg},
		Results:  results,
		Total:    len(results),
		Coverage: coverageFrom(pointer),
	}, nil
}

func dependencyWhyResults(
	graph depgraph.Graph,
	targets []dependencyTarget,
) []DependencyWhyResult {
	seen := map[string]DependencyWhyResult{}
	for _, target := range targets {
		for _, artifact := range graph.Artifacts {
			if !artifactMatchesTarget(artifact, target.coordinate) {
				continue
			}
			result := DependencyWhyResult{
				Coordinate:       artifact.Coordinate,
				Artifact:         artifact.Artifact,
				Version:          artifact.Version,
				Scope:            artifact.Scope,
				Direct:           artifact.Direct,
				Depth:            artifact.Depth,
				Path:             nonNil(artifact.Path),
				Parent:           artifact.Parent,
				DirectDependency: directDependencyFor(artifact),
				MatchedBy:        target.matchedBy,
				SourceJar:        target.sourceJar,
				SourceClass:      target.sourceClass,
			}
			key := result.Coordinate + "\x00" + result.MatchedBy + "\x00" + result.SourceClass
			seen[key] = result
		}
	}
	results := make([]DependencyWhyResult, 0, len(seen))
	for _, result := range seen {
		results = append(results, result)
	}
	sortDependencyWhy(results)
	return results
}

type dependencyTarget struct {
	coordinate  string
	matchedBy   string
	sourceJar   string
	sourceClass string
}

func dependencyTargets(
	ctx context.Context,
	pointer project.Pointer,
	sources []manifestSource,
	arg string,
) ([]dependencyTarget, error) {
	targets := []dependencyTarget{}
	if strings.Contains(arg, ".") && !strings.Contains(arg, ":") && !strings.HasSuffix(arg, ".jar") {
		classTargets, err := classDependencyTargets(ctx, pointer, sources, arg)
		if err != nil {
			return nil, err
		}
		targets = append(targets, classTargets...)
	}
	for _, source := range sources {
		switch {
		case source.Coordinate != "" &&
			(artifactMatchesArg(source.Coordinate, arg) || coordinateArtifactID(source.Coordinate) == arg):
			targets = append(targets, dependencyTarget{
				coordinate: source.Coordinate,
				matchedBy:  coordinateMatchReason(source.Coordinate, arg),
				sourceJar:  source.Source,
			})
		case source.Filename == arg || filepath.Base(source.Filename) == arg:
			targets = append(targets, dependencyTarget{
				coordinate: source.Coordinate,
				matchedBy:  "jar_filename",
				sourceJar:  source.Source,
			})
		}
	}
	if len(targets) == 0 && strings.Contains(arg, ":") {
		targets = append(targets, dependencyTarget{coordinate: arg, matchedBy: "coordinate"})
	}
	return dedupeDependencyTargets(targets), nil
}

func classDependencyTargets(
	ctx context.Context,
	pointer project.Pointer,
	sources []manifestSource,
	classFQN string,
) ([]dependencyTarget, error) {
	session, err := openReadOnly(pointer.SessionDBPath, true)
	if err != nil {
		return nil, fmt.Errorf("open session database: %w", err)
	}
	defer session.Close()
	rows, err := session.QueryContext(
		ctx,
		`SELECT source_shard_id FROM classes WHERE fqn = ? ORDER BY source_shard_id`,
		classFQN,
	)
	if err != nil {
		return nil, fmt.Errorf("query class source: %w", err)
	}
	defer rows.Close()
	byShard := map[string]manifestSource{}
	for _, source := range sources {
		byShard[source.ShardID] = source
	}
	targets := []dependencyTarget{}
	for rows.Next() {
		var shardID string
		if err := rows.Scan(&shardID); err != nil {
			return nil, fmt.Errorf("scan class source: %w", err)
		}
		source := byShard[shardID]
		if source.Coordinate == "" {
			continue
		}
		targets = append(targets, dependencyTarget{
			coordinate:  source.Coordinate,
			matchedBy:   "class",
			sourceJar:   source.Source,
			sourceClass: classFQN,
		})
	}
	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("iterate class sources: %w", err)
	}
	return targets, nil
}

func loadManifestSources(ctx context.Context, manifestPath string) ([]manifestSource, error) {
	rows, err := loadManifestRows(ctx, manifestPath)
	if err != nil {
		return nil, err
	}
	sources := make([]manifestSource, 0, len(rows))
	for _, row := range rows {
		sources = append(sources, manifestSource{
			ShardID:    row.shardID,
			Coordinate: row.coordinate,
			Filename:   row.filename,
			Source:     row.source,
		})
	}
	return sources, nil
}

func artifactMatchesTarget(artifact depgraph.Artifact, target string) bool {
	if target == "" {
		return false
	}
	return artifactMatchesArg(artifact.Coordinate, target) ||
		artifact.Artifact == target ||
		artifact.Name == target ||
		artifact.Coordinate == target
}

func artifactMatchesArg(coordinate, arg string) bool {
	if coordinate == arg {
		return true
	}
	parts := strings.Split(coordinate, ":")
	if len(parts) >= 2 && parts[0]+":"+parts[1] == arg {
		return true
	}
	return false
}

func coordinateArtifactID(coordinate string) string {
	parts := strings.Split(coordinate, ":")
	if len(parts) < 2 {
		return ""
	}
	return parts[1]
}

func coordinateMatchReason(coordinate, arg string) string {
	if coordinateArtifactID(coordinate) == arg {
		return "artifact"
	}
	return "coordinate"
}

func directDependencyFor(artifact depgraph.Artifact) string {
	if len(artifact.Path) > 0 {
		return artifact.Path[0]
	}
	if artifact.Direct {
		return artifact.Coordinate
	}
	return ""
}

func dedupeDependencyTargets(targets []dependencyTarget) []dependencyTarget {
	seen := map[string]bool{}
	result := []dependencyTarget{}
	for _, target := range targets {
		key := target.coordinate + "\x00" + target.matchedBy + "\x00" + target.sourceClass
		if seen[key] {
			continue
		}
		seen[key] = true
		result = append(result, target)
	}
	return result
}

func sortDependencyWhy(results []DependencyWhyResult) {
	sort.Slice(results, func(i, j int) bool {
		if results[i].Depth != results[j].Depth {
			return results[i].Depth < results[j].Depth
		}
		if results[i].Coordinate != results[j].Coordinate {
			return results[i].Coordinate < results[j].Coordinate
		}
		return results[i].MatchedBy < results[j].MatchedBy
	})
}

File: core/src/main/kotlin/dev/research4jar/query/Dependencies.kt
package dev.research4jar.query

import com.fasterxml.jackson.annotation.JsonInclude
import com.fasterxml.jackson.annotation.JsonProperty

/**
 * why-dependency: explains which resolved Maven dependency owns an argument
 * (class FQN, jar filename, artifact id, or coordinate). Ported from
 * querier/internal/query/dependencies.go; JSON key sets and SQL must stay
 * identical to the Go querier.
 */
data class DependencyWhyResult(
    @JsonProperty("coordinate") val coordinate: String,
    @JsonProperty("artifact") val artifact: String,
    @JsonProperty("version") val version: String,
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    @JsonProperty("scope") val scope: String = "",
    @JsonProperty("direct") val direct: Boolean,
    @JsonProperty("depth") val depth: Int,
    @JsonProperty("path") val path: List<String>,
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    @JsonProperty("parent") val parent: String = "",
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    @JsonProperty("direct_dependency") val directDependency: String = "",
    @JsonProperty("matched_by") val matchedBy: String,
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    @JsonProperty("source_jar") val sourceJar: String = "",
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    @JsonProperty("source_class") val sourceClass: String = "",
)

data class DependencyWhyResponse(
    @JsonProperty("query") val query: SymbolRequest,
    @JsonProperty("results") val results: List<DependencyWhyResult>,
    @JsonProperty("total") val total: Int,
    @JsonProperty("coverage") val coverage: Coverage,
)

fun whyDependency(
    pointer: ProjectPointerData,
    manifestPath: String,
    projectDir: String,
    arg: String,
): DependencyWhyResponse {
    val graph = try {
        DepGraphFile.load(projectDir)
    } catch (_: DepGraphUnsupportedException) {
        throw RuntimeException(
            "no dependency provenance found; run research4jar index in a Maven project" +
                " to create .research4jar/dependencies.json",
        )
    } catch (exception: Exception) {
        throw RuntimeException(
            "load dependency provenance: ${exception.message ?: exception}",
            exception,
        )
    }
    val sources = loadManifestSources(manifestPath)
    val targets = dependencyTargets(pointer, sources, arg)
    val results = dependencyWhyResults(graph, targets)
    return DependencyWhyResponse(
        query = SymbolRequest(command = "why-dependency", arg = arg),
        results = results,
        total = results.size,
        coverage = coverageFrom(pointer),
    )
}

// Go dependencyWhyResults: dedupes matches into a map keyed by
// coordinate/matchedBy/sourceClass, then sorts. Kotlin keeps insertion order
// for equal sort keys where Go map iteration is unordered; the comparator
// covers the full dedupe key minus sourceClass, so ties are rare.
internal fun dependencyWhyResults(
    graph: Graph,
    targets: List<DependencyTarget>,
): List<DependencyWhyResult> {
    val seen = LinkedHashMap<String, DependencyWhyResult>()
    for (target in targets) {
        for (artifact in graph.artifacts) {
            if (!artifactMatchesTarget(artifact, target.coordinate)) {
                continue
            }
            val result = DependencyWhyResult(
                coordinate = artifact.coordinate,
                artifact = artifact.artifact,
                version = artifact.version,
                scope = artifact.scope,
                direct = artifact.direct,
                depth = artifact.depth,
                // Go wraps with nonNil (configprops.go); the parsed
                // Artifact.path is already never null here.
                path = artifact.path,
                parent = artifact.parent,
                directDependency = directDependencyFor(artifact),
                matchedBy = target.matchedBy,
                sourceJar = target.sourceJar,
                sourceClass = target.sourceClass,
            )
            val key = result.coordinate + "\u0000" + result.matchedBy +
                "\u0000" + result.sourceClass
            seen[key] = result
        }
    }
    return seen.values.sortedWith(dependencyWhyComparator)
}

internal data class DependencyTarget(
    val coordinate: String,
    val matchedBy: String,
    val sourceJar: String = "",
    val sourceClass: String = "",
)

internal fun dependencyTargets(
    pointer: ProjectPointerData,
    sources: List<CachedManifestRow>,
    arg: String,
): List<DependencyTarget> {
    val targets = mutableListOf<DependencyTarget>()
    if (arg.contains(".") && !arg.contains(":") && !arg.endsWith(".jar")) {
        targets += classDependencyTargets(pointer, sources, arg)
    }
    for (source in sources) {
        when {
            source.coordinate.isNotEmpty() &&
                (
                    artifactMatchesArg(source.coordinate, arg) ||
                        coordinateArtifactId(source.coordinate) == arg
                    ) -> {
                targets += DependencyTarget(
                    coordinate = source.coordinate,
                    matchedBy = coordinateMatchReason(source.coordinate, arg),
                    sourceJar = source.source,
                )
            }
            source.filename == arg || filepathBase(source.filename) == arg -> {
                targets += DependencyTarget(
                    coordinate = source.coordinate,
                    matchedBy = "jar_filename",
                    sourceJar = source.source,
                )
            }
        }
    }
    if (targets.isEmpty() && arg.contains(":")) {
        targets += DependencyTarget(coordinate = arg, matchedBy = "coordinate")
    }
    return dedupeDependencyTargets(targets)
}

private fun classDependencyTargets(
    pointer: ProjectPointerData,
    sources: List<CachedManifestRow>,
    classFqn: String,
): List<DependencyTarget> = Db.openReadOnly(pointer.sessionDbPath, immutable = true).use { session ->
    val shardIds = session.query(
        "SELECT source_shard_id FROM classes WHERE fqn = ? ORDER BY source_shard_id",
        listOf(classFqn),
    ) { rows -> rows.mapRows { it.getString(1) } }
    val byShard = ManifestCache.mapSessionRows(session, sources, shardIds)
    val targets = mutableListOf<DependencyTarget>()
    for (shardId in shardIds) {
        val source = byShard[shardId] ?: continue
        if (source.coordinate.isEmpty()) {
            continue
        }
        targets += DependencyTarget(
            coordinate = source.coordinate,
            matchedBy = "class",
            sourceJar = source.source,
            sourceClass = classFqn,
        )
    }
    targets
}

/** Go manifestSource carries the same four columns as [CachedManifestRow]. */
internal fun loadManifestSources(manifestPath: String): List<CachedManifestRow> =
    ManifestCache.loadRows(manifestPath)

private fun artifactMatchesTarget(artifact: Artifact, target: String): Boolean {
    if (target.isEmpty()) {
        return false
    }
    return artifactMatchesArg(artifact.coordinate, target) ||
        artifact.artifact == target ||
        artifact.name == target ||
        artifact.coordinate == target
}

internal fun artifactMatchesArg(coordinate: String, arg: String): Boolean {
    if (coordinate == arg) {
        return true
    }
    val parts = coordinate.split(":")
    return parts.size >= 2 && parts[0] + ":" + parts[1] == arg
}

internal fun coordinateArtifactId(coordinate: String): String {
    val parts = coordinate.split(":")
    if (parts.size < 2) {
        return ""
    }
    return parts[1]
}

private fun coordinateMatchReason(coordinate: String, arg: String): String =
    if (coordinateArtifactId(coordinate) == arg) "artifact" else "coordinate"

private fun directDependencyFor(artifact: Artifact): String {
    if (artifact.path.isNotEmpty()) {
        return artifact.path[0]
    }
    if (artifact.direct) {
        return artifact.coordinate
    }
    return ""
}

internal fun dedupeDependencyTargets(targets: List<DependencyTarget>): List<DependencyTarget> {
    val seen = mutableSetOf<String>()
    val result = mutableListOf<DependencyTarget>()
    for (target in targets) {
        val key = target.coordinate + "\u0000" + target.matchedBy + "\u0000" + target.sourceClass
        if (!seen.add(key)) {
            continue
        }
        result += target
    }
    return result
}

private val dependencyWhyComparator =
    compareBy<DependencyWhyResult>({ it.depth }, { it.coordinate }, { it.matchedBy })

/**
 * Go filepath.Base for slash-separated paths: manifest jar filenames and walk
 * paths on the platforms the e2e suite runs on. "" maps to "." exactly as in
 * Go, which downstream term lists rely on.
 */
internal fun filepathBase(path: String): String {
    if (path.isEmpty()) {
        return "."
    }
    var trimmed = path
    while (trimmed.isNotEmpty() && trimmed.endsWith('/')) {
        trimmed = trimmed.dropLast(1)
    }
    val index = trimmed.lastIndexOf('/')
    val base = if (index >= 0) trimmed.substring(index + 1) else trimmed
    return if (base.isEmpty()) "/" else base
}
